lib/blockstore/file.js

```diff
@@ -0,0 +1,579 @@
+
+    const magic = bwm.render();
+
+    const {
+      fileno,
+      filerecord,
+      filepath
+    } = await this.allocate(type, length);
+
+    const mposition = filerecord.used;
+    const bposition = filerecord.used + mlength;
+
+    const fd = await fs.open(filepath, 'r+');
+
+    const mwritten = await fs.write(fd, magic, 0, mlength, mposition);
+    const bwritten = await fs.write(fd, data, 0, blength, bposition);
+
+    await fs.close(fd);
+
+    if (mwritten !== mlength) {
+      this.writing = false;
+      throw new Error('Could not write block magic.');
+    }
+
+    if (bwritten !== blength) {
+      this.writing = false;
+      throw new Error('Could not write block.');
+    }
+
+    filerecord.blocks += 1;
+    filerecord.used += length;
+
+    const b = this.db.batch();
+
+    const blockrecord = new BlockRecord({
+      file: fileno,
+      position: bposition,
+      length: blength
+    });
+
+    b.put(layout.b.encode(type, hash), blockrecord.toRaw());
+    b.put(layout.f.encode(type, fileno), filerecord.toRaw());
+
+    const last = bio.write(4).writeU32(fileno).render();
+    b.put(layout.F.encode(type), last);
+
+    await b.write();
+
+    this.writing = false;
+
+    return true;
+  }
+
+  /**
+   * This method will retrieve block undo coin data.
+   * @param {Buffer} hash - The block hash
+   * @returns {Promise}
+   */
+
+  async readUndo(hash) {
+    return this._read(types.UNDO, hash);
+  }
+
+  /**
+   * This method will retrieve block data. Smaller portions of the
+   * block (e.g. transactions) can be read by using the offset and
+   * length arguments.
+   * @param {Buffer} hash - The block hash
+   * @param {Number} offset - The offset within the block
+   * @param {Number} length - The number of bytes of the data
+   * @returns {Promise}
+   */
+
+  async read(hash, offset, length) {
+    return this._read(types.BLOCK, hash, offset, length);
+  }
+
+  /**
+   * This methods reads data from disk by retrieving the index of
+   * the data and reading from the corresponding file and location.
+   * @private
+   * @param {Number} type - The type of block data
+   * @param {Buffer} hash - The block hash
+   * @param {Number} offset - The offset within the block
+   * @param {Number} length - The number of bytes of the data
+   * @returns {Promise}
+   */
+
+  async _read(type, hash, offset, length) {
+    const raw = await this.db.get(layout.b.encode(type, hash));
+    if (!raw)
+      return null;
+
+    const blockrecord = BlockRecord.fromRaw(raw);
+
+    const filepath = this.filepath(type, blockrecord.file);
+
+    let position = blockrecord.position;
+
+    if (offset)
+      position += offset;
+
+    if (!length)
+      length = blockrecord.length;
+
+    if (offset + length > blockrecord.length)
+      throw new Error('Out-of-bounds read.');
+
+    const data = Buffer.alloc(length);
+
+    const fd = await fs.open(filepath, 'r');
+    await fs.read(fd, data, 0, length, position);
```

Comment:
Should also be wrapped in a `try...finally` block

Reply:
Added